Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+        }
+        
+        let stream = HTTP2StreamData.getStream(for: streamID, allocator: self.allocator)
+        
+        // this just contains a header block fragment and flags
+        let headers: HPACKHeaders
+        if bytes.readableBytes == length {
+            headers = try stream.decoder.decodeHeaders(from: &bytes)
+        } else {
+            var slice = bytes.readSlice(length: length)!
+            headers = try stream.decoder.decodeHeaders(from: &slice)
+        }
+        
+        switch frame.payload {
+        case .headers(var currentHeaders, let priority):
+            currentHeaders.addContinuation(headers)
+            frame.payload = .headers(currentHeaders, priority)
+            
+        case .pushPromise(let streamID, var currentHeaders):
+            currentHeaders.addContinuation(headers)
+            frame.payload = .pushPromise(streamID, currentHeaders)
+            
+        default:
+            // A CONTINUATION frame MUST be preceded by a HEADERS, PUSH_PROMISE or
+            // CONTINUATION frame without the END_HEADERS flag set. A recipient that
+            // observes violation of this rule MUST respond with a connection error
+            // (Section 5.4.1) of type PROTOCOL_ERROR.
+            throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+        }
+        
+        if flags.contains(.endHeaders) {
+            // no continuations following, so emit a full frame
+            frame.flags.formUnion(.endHeaders)
+            self.currentHeaderFrame = nil
+            return frame
+        } else {
+            self.currentHeaderFrame = frame
+            return nil
+        }
+    }
+}
+
+class HTTP2FrameEncoder {
+    private let allocator: ByteBufferAllocator
+    
+    init(allocator: ByteBufferAllocator) {
+        self.allocator = allocator
+    }
+    
+    /// Encodes the frame and optionally returns one or more blobs of data
+    /// ready for the system.
+    ///
+    /// Returned data blobs would include anything of potentially flexible
+    /// length, such as DATA payloads, header fragments in HEADERS or PUSH_PROMISE
+    /// frames, and so on. This is to avoid manually copying chunks of data which
+    /// we could just enqueue separately in sequence on the channel. Generally, if
+    /// we have a byte buffer somewhere, we will return that separately rather than
+    /// copy it into another buffer, with the corresponding allocation overhead.
+    ///
+    /// - Parameters:
+    ///   - frame: The frame to encode.
+    ///   - buf: Destination buffer for the encoded frame.
+    /// - Returns: An array containing zero or more additional buffers to send, in
+    ///            order. These may contain data frames' payload bytes, encoded
+    ///            header fragments, etc.
+    /// - Throws: Errors returned from HPACK encoder.
+    func encode(frame: HTTP2Frame, to buf: inout ByteBuffer) throws -> [IOData] {
```

Comment:
Will the returned blobs ever have more than two elements? If not, it's better to return these as a tuple than as an array, to avoid the heap allocation to store the returned elements.